### PR TITLE
Use generator resume arguments in the async/await lowering

### LINKED ISSUE #56974: Cannot define async fn with no_std
```rust
#![feature(async_await, futures_api)]

#![no_std]

pub async fn foo() {
}
```

([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ff1a2fae4bc9507da9489c5fc1315576)) currently fails with 

```
error[E0433]: failed to resolve: could not find `from_generator` in `future`
```

This should also work with whatever the syntax for `await` is, currently it fails earlier than this error because of `await!` only being defined in `std`, see #56767.

### PR DESCRIPTION
**Pull request by jonas-schievink**
This removes the TLS requirement from async/await and enables it in `#![no_std]` crates.

Closes https://github.com/rust-lang/rust/issues/56974

I'm not confident the HIR lowering is completely correct, there seem to be quite a few undocumented invariants in there. The `async-std` and tokio test suites are passing with these changes though.

### CODE REVIEW

File: src/librustc_ast_lowering/expr.rs [hunk 1]
```diff
@@ -523,9 +554,10 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
-    ///         match ::std::future::poll_with_tls_context(unsafe {
-    ///             <::std::pin::Pin>::new_unchecked(&mut pinned)
-    ///         }) {
+    ///         match unsafe { ::std::future::poll_with_context(
+    ///             <::std::pin::Pin>::new_unchecked(&mut pinned),
+    ///             task_context,
+    ///         ) } {
     ///             ::std::task::Poll::Ready(result) => break result,
     ///             ::std::task::Poll::Pending => {}
     ///         }
```
Comment (Arnavion) [inline src/librustc_ast_lowering/expr.rs]: Github won't let me add a comment on line 564 itself, but shouldn't that be `task_context = yield ();` now?

File: src/librustc_ast_lowering/expr.rs [hunk 2]
```diff
@@ -629,12 +671,23 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
+                // Use of `await` outside of an async context. Return `yield_expr` so that we can
+                // proceed with type checking.
+                self.stmt(span, hir::StmtKind::Semi(yield_expr))
+            }
         };
 
         let loop_block = self.block_all(span, arena_vec![self; inner_match_stmt, yield_stmt], None);
 
-        // loop { .. }
+        // loop { ...; task_context = yield (); }
```
Comment (Arnavion) [inline src/librustc_ast_lowering/expr.rs]: Should this `task_context = yield ();` addition instead be a comment on line 678, which is where the assignment is generated?

File: src/librustc_ast_lowering/expr.rs (hunk 1 above)
Comment (jonas-schievink) [reply src/librustc_ast_lowering/expr.rs]: Yeah, good catch!

File: src/librustc_ast_lowering/expr.rs (hunk 2 above)
Comment (jonas-schievink) [reply src/librustc_ast_lowering/expr.rs]: Yes, that makes sense to me

File: src/librustc_ast_lowering/expr.rs [hunk 3]
```diff
@@ -483,14 +484,44 @@ impl<'hir> LoweringContext<'_, 'hir> {
             Some(ty) => FunctionRetTy::Ty(ty),
             None => FunctionRetTy::Default(span),
         };
-        let ast_decl = FnDecl { inputs: vec![], output };
+
+        let task_context_id = self.resolver.next_node_id();
+        let task_context_hid = self.lower_node_id(task_context_id);
+        let ast_decl = FnDecl {
```
Comment (Centril) [inline src/librustc_ast_lowering/expr.rs]: Style: lowering code is typically written using more `let` bindings to break up large expressions like this.

File: src/librustc_ast_lowering/expr.rs [hunk 4]
```diff
@@ -483,14 +484,44 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
+                        None,
+                    ),
+                    span: DUMMY_SP,
+                }),
+                id: self.resolver.next_node_id(),
+                span: DUMMY_SP,
+                is_placeholder: false,
+            }],
+            output,
+        };
```
Comment (Centril) [inline src/librustc_ast_lowering/expr.rs]: I don't think we should be creating `syntax::ast` IR inside lowering just to immediately convert it to HIR. Instead, the HIR representation should be created directly. (cc @eddyb)

File: src/libcore/future/mod.rs [hunk 5]
```diff
@@ -2,6 +2,84 @@
...
+/// better error messages (`impl Future` rather than `GenFuture<[closure.....]>`).
+#[doc(hidden)]
+#[unstable(feature = "gen_future", issue = "50547")]
+#[cfg(not(bootstrap))]
+#[inline]
+pub fn from_generator<T>(gen: T) -> impl Future<Output = T::Return>
+where
+    T: Generator<ResumeTy, Yield = ()>,
+{
+    #[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
```
Comment (Centril) [inline src/libcore/future/mod.rs]: Are all of these actually used?

File: src/test/ui/async-await/no-std.rs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+#![no_std]
+#![crate_type = "rlib"]
+
+use core::future::Future;
+
+async fn a(f: impl Future) {
+    f.await;
+}
+
+fn main() {}
```
Comment (Centril) [inline src/test/ui/async-await/no-std.rs]: Could we test more like a whole program with more interactions?

**Comment by Centril** [comment]
cc https://github.com/rust-lang/rust/issues/69039 -- should be fixed before this can land?

**Comment by jonas-schievink** [comment]
> cc #69039 -- should be fixed before this can land?

That seems to be an issue with drops – async/await only needs to be able to resume with `Copy` data.

**Comment by jonas-schievink** [comment]
Ah, no, it applies whenever data is stored across a `yield`, but we don't do that here as far as I can tell.

**Comment by jonas-schievink** [comment]
Ah, no, I'll have to take a deeper look at that

**Comment by jonas-schievink** [comment]
#69039 should only be an issue if the resume argument is stored across `yield` points, and we don't do that here

File: src/librustc_ast_lowering/expr.rs (hunk 4 above)
Comment (jonas-schievink) [reply src/librustc_ast_lowering/expr.rs]: That makes sense to me, but there seems to be no existing code that directly creates an `hir::FnDecl`, and `lower_fn_decl` does quite a lot of work that I don't really want to hand-roll.

Comment (Centril) [reply src/librustc_ast_lowering/expr.rs]: It's not that much code tbh which I think can be mostly shared by some refactorings of `lower_fn_decl` into the main parts (inputs, output, implicit_self, packaging those aforementioned up) as separate methods.

**Comment by nikomatsakis** [comment]
I'm excited to see this, thanks @jonas-schievink

**Comment by CryZe** [comment]
What's the status of this? Is this blocked on #69039 or could this be merged before this is resolved?

**Comment by jonas-schievink** [comment]
@CryZe I've started to write a test that should provide an answer to that, but I also plan to look into #69039 regardless. I hope to find some time to finish this off in the next couple of weeks.

File: src/libcore/future/mod.rs [hunk 7]
```diff
@@ -2,6 +2,84 @@
...
+    task::{Context, Poll},
+};
+
 mod future;
 #[stable(feature = "futures_api", since = "1.36.0")]
 pub use self::future::Future;
+
+/// This type is needed because:
+///
+/// a) Generators cannot implement `for<'a, 'b> Generator<&'a mut Context<'b>>`, so we need to pass
```
Comment (tmandry) [inline src/libcore/future/mod.rs]: Why not? `impl<'a, 'b> Generator<&'a mut Context<'b>> for ...`

**Review by tmandry — CHANGES_REQUESTED** [review]

File: src/libcore/future/mod.rs (hunk 7 above)
Comment (jonas-schievink) [reply src/libcore/future/mod.rs]: Generator literals don't support that at the moment, see https://github.com/rust-lang/rust/issues/68923

File: src/librustc_span/symbol.rs [hunk 8]
```diff
@@ -715,6 +715,7 @@ symbols! {
         target_has_atomic_load_store,
         target_thread_local,
         task,
+        _task_context,
```
[the remark this answers is not in the document]
Comment (jonas-schievink) [reply src/librustc_span/symbol.rs]: It's there because it suppresses the unused lint. I found that much easier than trying to create an `#[allow(unused)]` attribute in the HIR lowering code, but can do that if it's preferred.

[the remark this answers is not in the document]
Comment (tmandry) [reply src/librustc_span/symbol.rs]: oh, I don't understand why the lint fires when you reference `sym::_task_context` elsewhere in the code. I was assuming you had it to suppress the lint at first, but didn't need it anymore.

File: src/libcore/future/mod.rs [hunk 9]
```diff
@@ -2,6 +2,84 @@
...
+
+    impl<T: Generator<ResumeTy, Yield = ()>> Future for GenFuture<T> {
+        type Output = T::Return;
+        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            // Safety: Safe because we're !Unpin + !Drop mapping to a ?Unpin value
+            let gen = unsafe { Pin::map_unchecked_mut(self, |s| &mut s.0) };
+
+            // Resume the generator, turning the `&mut Context` into a `NonNull` raw pointer. The
+            // `.await` lowering will safely cast that back to a `&mut Context`.
+            match gen.resume(ResumeTy(NonNull::from(cx).cast::<Context<'static>>())) {
```
[the remark this answers is not in the document]
Comment (jonas-schievink) [reply src/libcore/future/mod.rs]: I'm not sure what `poll_fn` is in this example, it doesn't seem to exist.

My idea was that the only way to get the waker back out of the generator is via `poll_with_context` below, which only calls `Future::poll`. That only takes a `&mut Context` with a fully generic lifetime, so at that point the `'static` lifetime gets demoted back to `'a`.

File: src/libcore/future/mod.rs [hunk 10]
```diff
@@ -2,6 +2,83 @@
...
+///    a raw pointer.
+/// b) Raw pointers and `NonNull` aren't `Send` or `Sync`, so that would make every single future
+///    non-Send/Sync as well, and we don't want that.
+///
+/// It also simplifies the HIR lowering of `.await`.
+#[doc(hidden)]
+#[unstable(feature = "gen_future", issue = "50547")]
+#[cfg(not(bootstrap))]
+#[derive(Debug, Copy, Clone)]
+pub struct ResumeTy(pub NonNull<Context<'static>>);
```
Comment (tmandry) [inline src/libcore/future/mod.rs]: Can we make the inner field private? I don't see any benefit of making it public.

File: src/libcore/future/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tmandry) [reply src/libcore/future/mod.rs]: Ah, I understand now. Thanks.

`cx.0.as_mut()` below does create a `&mut Context<'static>` _in the context of that function_, which means the compiler could observe that there exists a `&'static Waker`, pointing to some memory which is not actually valid for the static lifetime. Is there any way that can create UB? cc @RalfJung

File: src/libcore/future/mod.rs [hunk 11]
```diff
@@ -2,6 +2,83 @@
...
+#[doc(hidden)]
+#[unstable(feature = "gen_future", issue = "50547")]
+#[cfg(not(bootstrap))]
+#[inline]
+pub unsafe fn poll_with_context<F>(f: Pin<&mut F>, mut cx: ResumeTy) -> Poll<F::Output>
+where
+    F: Future,
+{
+    F::poll(f, cx.0.as_mut())
+}
```
Comment (tmandry) [inline src/libcore/future/mod.rs]: We could avoid the above situation (if it is indeed a problem) by doing something like the following. `'a` is a pessimistic bound on the inner lifetime of `Context`, which is fine.

```rust
pub unsafe fn poll_with_context<'a, F>(f: Pin<&mut F>, &'a mut cx: ResumeTy) -> Poll<F::Output>
where
    F: Future,
{
    let cx: NonNull<Context<'a>> = cx.0.cast();
    F::poll(f, cx.as_mut())
}
```

**Review by tmandry — COMMENTED** [review]
Sorry for the latency on my end.

File: src/libcore/future/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/libcore/future/mod.rs]: I'm entirely lacking context here, so it's hard to say anything specific.
But: one design goal of Stacked Borrows (and for now I intend to maintain that for as we keep evolving our aliasing model) is that lifetimes don't matter. If the *only* difference between two programs, at the MIR level, is their lifetimes, then they will behave the same also in terms of UB. Stacked Borrows care about when a reference is used (where "use" includes all forms of dereferencing, in particular reborrowing), but it doesn't care if the lifetime goes on way after the last use of a reference.

In other words, this function is totally fine for Stacked Borrows:
```rust
pub fn write(x: &mut i32, val: i32) {
    let static_x: &'static mut i32 = unsafe { std::mem::transmute(x) };
    *static_x = val;
}
```

Of course, differences in lifetimes could lead to rustc picking different trait implementations or some such thing, and that *will* affect behavior.

[the remark this answers is not in the document]
Comment (tmandry) [reply src/libcore/future/mod.rs]: Okay, it sounds like this isn't a problem in Stacked Borrows then. Thanks!

File: src/librustc_ast_lowering/expr.rs (hunk 4 above)
Comment (jonas-schievink) [reply src/librustc_ast_lowering/expr.rs]: I've pushed a commit that should address this

File: src/librustc_span/symbol.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jonas-schievink) [reply src/librustc_span/symbol.rs]: Oh, it doesn't fire on the compiler, if that's what you mean. It fires on the code being compiled when it doesn't `.await` anything IIRC.

**Comment by jonas-schievink** [comment]
#69716 has landed now, and the size tests are passing again

File: src/libcore/future/mod.rs (hunk 7 above)
Comment (tmandry) [reply src/libcore/future/mod.rs]: Can you leave a pointer to #68923 here?

**Comment by tmandry** [comment]
Looks good, this is exciting!

r=me after final [nit](https://github.com/rust-lang/rust/pull/69033#discussion_r393903157)

File: src/test/ui/async-await/no-std.rs (hunk 6 above)
Comment (vertexclique) [reply src/test/ui/async-await/no-std.rs]: Please add more run-pass with:
* spawn onto the main thread with a basic waker vtable
* a spinlock holds that task

Comment (jonas-schievink) [reply src/test/ui/async-await/no-std.rs]: There are already a lot of tests for async/await, most just don't turn on the `#![no_std]` flag. Since the underlying implementation is the same there is not really a difference though.

(I'm not entirely sure what the spinlock test would accomplish here though, you can store tasks however you want as long as you can get a `Pin<&mut Self>` to poll the future)

Comment (vertexclique) [reply src/test/ui/async-await/no-std.rs]: > There are already a lot of tests for async/await, most just don't turn on the #![no_std] flag.

If we have a build matrix that enables no_std flag for async/await tests, it is ok to not add tests.(iirc we don't have) I assume compiler harness doesn't do that unless you explicitly ask for it. So when it is passing for std doesn't mean that it will pass by all cases of no_std.

> (I'm not entirely sure what the spinlock test would accomplish here though,

To demonstrate also it is working in no_std env without any extra dependency brought in.

File: src/librustc_ast_lowering/expr.rs [hunk 12]
```diff
@@ -480,17 +489,42 @@ impl<'hir> LoweringContext<'_, 'hir> {
...
+
+            let old_ctx = this.task_context;
+            this.task_context = Some(task_context_hid);
+            let res = body(this);
+            this.task_context = old_ctx;
+            (params, res)
         });
 
-        // `static || -> <ret_ty> { body }`:
+        // `static |_task_context| -> <ret_ty> { body }`:
```
Comment (RalfJung) [inline src/librustc_ast_lowering/expr.rs]: Reading just the comments, I am a bit confused by the leading `_` here -- shouldn't this be the binder for the `task_context` below?

**Comment by Centril** [comment]
The language team hasn't had a chance to discuss and approve this PR yet, which I fully expect we will do, (I'm also excited to see this land), but the process for language changes (specifically the new internal dependence on resume arguments) should be followed. We should have time to rubber-stamp this on our meeting tomorrow, but for now, @bors r-

File: src/librustc_ast_lowering/expr.rs (hunk 12 above)
Comment (jonas-schievink) [reply src/librustc_ast_lowering/expr.rs]: The variable in the desugaring has an underscore because it can trigger the unused variable lint otherwise. I didn't mirror that in the lowering code.

Comment (RalfJung) [reply src/librustc_ast_lowering/expr.rs]: Ah so it should be `_task_context` in both places? Makes sense, thanks.

**Comment by Centril** [comment]
We discussed this on the language team meeting just a few minutes ago. Everyone was quite excited to rubber-stamp this and see it land. :)

@bors r=tmandry

🚀

**Comment by benma** [comment]
Tested in on nightly, works great. Awesome work!

A bit new to Rust release scheduling; will this be in `stable` with the next release, or do we need to wait until 1.44 is stable, which is not until June?

**Comment by jonas-schievink** [comment]
This will land in 1.44 (and will hit beta in 1.43)

**Comment by benma** [comment]
Any chance to put it into the next stable release? After all, it's an implementation detail, not a language change.

**Comment by jonas-schievink** [comment]
No, we only backport bug fixes

**Comment by kpp** [comment]
@jonas-schievink Hi. I got a 300% performance degradation on 1.44.0-nightly 2fbb07525 2020-03-26 compared to stable 1.42 b8cedc004 2020-03-09 while [benching](https://github.com/kpp/futures-async-combinators/) futures.

<details><summary>Criterion output</summary>
<p>

```
future::ready/futures   time:   [1.5667 ns 1.5712 ns 1.5762 ns]                                   
                        change: [+223.24% +232.94% +242.84%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
future::ready/async_combinators                                                                             
                        time:   [1.5699 ns 1.5743 ns 1.5793 ns]
                        change: [+279.32% +282.34% +285.20%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) low mild
  2 (2.00%) high mild
  3 (3.00%) high severe

future::poll_fn/futures time:   [260.26 ps 260.82 ps 261.40 ps]                                    
                        change: [-3.2265% -2.6712% -2.0788%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  5 (5.00%) high mild
  4 (4.00%) high severe
Benchmarking future::poll_fn/async_combinators: Collecting 100 samples in estimated 5.0000 s (19B iterations                                                                                                            future::poll_fn/async_combinators                        
                        time:   [260.18 ps 260.91 ps 261.77 ps]
                        change: [-1.3517% -0.6763% -0.0422%] (p = 0.05 < 0.05)
                        Change within noise threshold.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) low mild
  4 (4.00%) high mild
  3 (3.00%) high severe

future::map/futures     time:   [2.1487 ns 2.1544 ns 2.1601 ns]                                 
                        change: [+296.27% +298.58% +300.54%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
  4 (4.00%) high mild
  1 (1.00%) high severe
future::map/async_combinators                                                                             
                        time:   [3.1330 ns 3.1393 ns 3.1458 ns]
                        change: [+233.70% +236.33% +238.68%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) low mild
  3 (3.00%) high mild

     Running target/release/deps/stream-ff2c5e87168d8d4c
stream::iter/futures    time:   [3.1216 ns 3.1275 ns 3.1341 ns]                                  
                        change: [+290.24% +293.62% +297.39%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low mild
  6 (6.00%) high mild
  4 (4.00%) high severe
stream::iter/async_combinators                                                                             
                        time:   [3.1331 ns 3.1423 ns 3.1532 ns]
                        change: [+280.94% +283.80% +286.95%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe

stream::next/futures    time:   [3.3549 ns 3.4282 ns 3.5139 ns]                                  
                        change: [+319.16% +327.38% +336.10%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe
stream::next/async_combinators                                                                             
                        time:   [3.1806 ns 3.2330 ns 3.2933 ns]
                        change: [+288.17% +292.97% +297.98%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe

stream::collect/futures time:   [6.3038 ns 6.3446 ns 6.3928 ns]                                     
                        change: [+298.95% +301.74% +305.36%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 17 outliers among 100 measurements (17.00%)
  4 (4.00%) low mild
  9 (9.00%) high mild
  4 (4.00%) high severe
Benchmarking stream::collect/async_combinators: Collecting 100 samples in estimated 5.0000 s (456M iteration                                                                                                            stream::collect/async_combinators                        
                        time:   [11.661 ns 12.051 ns 12.435 ns]
                        change: [+320.32% +329.63% +340.16%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

stream::map/futures     time:   [6.8157 ns 6.9007 ns 6.9814 ns]                                 
                        change: [+305.37% +309.91% +314.44%] (p = 0.00 < 0.05)
                        Performance has regressed.
stream::map/async_combinators                                                                             
                        time:   [7.0993 ns 7.2319 ns 7.3715 ns]
                        change: [+339.77% +348.71% +358.18%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

stream::fold/futures    time:   [4.3133 ns 4.3516 ns 4.4018 ns]                                  
                        change: [+277.61% +283.58% +291.19%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
stream::fold/async_combinators                                                                             
                        time:   [9.3900 ns 9.4145 ns 9.4431 ns]
                        change: [+259.06% +261.65% +264.17%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  7 (7.00%) high mild
  6 (6.00%) high severe
```

</p>
</details>

I know my benchmarks are not perfect. Still 300% is a lot.

Would you please take a look?

**Comment by goffrie** [comment]
I was able to reproduce the regression, but it's not from this PR: 5f13820478907b09d50baf74f3ff2b78499ecd6c and 1057dc97afce39ff6a224966ece3ed438af4c1f5 are both equally slow.

**Comment by goffrie** [comment]
Bisecting shows that 57e1da59cd0761330b4ea8d47b16340a78eeafa9 introduced the regression.

**Comment by Mark-Simulacrum** [comment]
Please file new issue(s) for problems arising as a result of PRs, otherwise it's highly likely to get lost.
